### ISSUE TITLE
failed to call webhook despite the webhook exists - validate-metallb-io-v1beta1-ipaddresspool

### ISSUE THREAD

**Issue by CallisteH** [metallb/metallb#2486]
### TL;DR
I’ve installed MetalLB, but I’m unable to create an IPAddressPool because the validation webhook isn’t accessible. However, the validation webhook does exist, and I can query it using curl from a pod within the same namespace.

---
### Description 
I'm attempting to install MetalLB on my bare-metal Kubernetes cluster using the following command:

```bash
kubectl apply -f https://raw.githubusercontent.com/metallb/metallb/v0.13.12/config/manifests/metallb-native.yaml
```

However, when I try to apply the `ipaddresspool`, I encounter this error:

```bash
local@edge-master:~/k8s_manifests/admin/metallb$ kubectl apply -f ipaddresspool.yaml 
Error from server (InternalError): error when creating "ipaddresspool.yaml": Internal error occurred: failed calling webhook "ipaddresspoolvalidationwebhook.metallb.io": failed to call webhook: Post "https://webhook-service.metallb-system.svc:443/validate-metallb-io-v1beta1-ipaddresspool?timeout=10s": context deadline exceeded
```
Here is the IPAddressPool : 
```yaml
apiVersion: metallb.io/v1beta1
kind: IPAddressPool
metadata:
  name: address-pool
  namespace: metallb-system 
spec:
  addresses:
  - 192.168.1.1-192.168.1.254
```

The `ValidatingWebhookConfiguration` is present:

```bash
$ kubectl -n metallb-system get ValidatingWebhookConfiguration
NAME                                                   WEBHOOKS   AGE
metallb-webhook-configuration                          7          178m
```

The service also exists:

```bash
$ kubectl -n metallb-system get svc
NAME              TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)   AGE
webhook-service   ClusterIP   <ip>             <none>        443/TCP   179m
```

It seems that the webhook is accessible, at least from a pod in the same namespace. I created a debug pod with tools like `curl` and was able to reach the webhook:

```bash
curl https://webhook-service.metallb-system.svc:443/validate-metallb-io-v1beta1-ipaddresspool?timeout=10s -k
{"response":{"uid":"","allowed":false,"status":{"metadata":{},"message":"contentType=, expected application/json","code":400}}}
```

The response is a 400 error because the data and request type are incorrect (GET instead of POST), BUT the metallb-controller answers.

I'm struggling to debug this further. Could this be related to a problem with the CNI? I'm using Cilium.

### Expected Behavior

We should have the ipaddresspool object should be created correctly

### Additional Context

I'm not sure if this could be causing issues, but I'm experiencing errors with some of the Cilium pods.

I'm running Kubernetes with one master and two workers. I have four Cilium pods, but one of them is showing errors:

```bash
$ kubectl -n metallb-system logs cilium-mvmb2
...
level=error msg="cannot extract endpoint ID from DNS request" DNSRequestID=38072 dnsName=eviden.com. error="Cannot find endpoint with IP 172.8.8.3" ipAddr="172.8.8.3:36245" subsys=fqdn/dnsproxy
level=error msg="cannot find matching endpoint" error="DNS request cannot be associated with an existing endpoint" l3n4Addr="172.8.8.3:36245" subsys=daemon
level=error msg="cannot extract endpoint ID from DNS request" DNSRequestID=27306 dnsName=eviden.com. error="Cannot find endpoint with IP 172.8.8.3" ipAddr="172.8.8.3:36773" subsys=fqdn/dnsproxy
level=error msg="cannot find matching endpoint" error="DNS request cannot be associated with an existing endpoint" l3n4Addr="172.8.8.3:36773" subsys=daemon
level=error msg="cannot extract endpoint ID from DNS request" DNSRequestID=46019 dnsName=126.1.0.10.in-addr.arpa. error="Cannot find endpoint with IP 172.8.8.3" ipAddr="172.8.8.3:36773" subsys=fqdn/dnsproxy
level=error msg="cannot find matching endpoint" error="DNS request cannot be associated with an existing endpoint" l3n4Addr="172.8.8.3:36773" subsys=daemon
level=error msg="cannot extract endpoint ID from DNS request" DNSRequestID=26501 dnsName=126.1.0.10.in-addr.arpa. error="Cannot find endpoint with IP 172.8.8.3" ipAddr="172.8.8.3:36245" subsys=fqdn/dnsproxy
level=error msg="cannot find matching endpoint" error="DNS request cannot be associated with an existing endpoint" l3n4Addr="172.8.8.3:36245" subsys=daemon
level=warning msg="WriteMsgIP failed" error="write ip 127.0.0.1->0.0.0.0: sendmsg: invalid argument" subsys=fqdn/dnsproxy
level=warning msg="WriteMsgIP failed" error="write ip 127.0.0.1->172.7.2.101: sendmsg: invalid argument" subsys=fqdn/dnsproxy
level=warning msg="WriteMsgIP failed" error="write ip 127.0.0.1->10.22.80.5: sendmsg: invalid argument" subsys=fqdn/dnsproxy
level=error msg="cannot extract endpoint ID from DNS request" DNSRequestID=35178 dnsName=3.ubuntu.pool.ntp.org.seed. error="Cannot find endpoint with IP 127.0.0.1" ipAddr="127.0.0.1:37027" subsys=fqdn/dnsproxy
level=error msg="cannot extract endpoint ID from DNS request" DNSRequestID=60256 dnsName=3.ubuntu.pool.ntp.org.seed. error="Cannot find endpoint with IP 127.0.0.1" ipAddr="127.0.0.1:37027" subsys=fqdn/dnsproxy
level=error msg="cannot find matching endpoint" error="DNS request cannot be associated with an existing endpoint" l3n4Addr="127.0.0.1:37027" subsys=daemon
level=error msg="cannot find matching endpoint" error="DNS request cannot be associated with an existing endpoint" l3n4Addr="127.0.0.1:37027" subsys=daemon
level=error msg="cannot extract endpoint ID from DNS request" DNSRequestID=50573 dnsName=126.1.0.10.in-addr.arpa. error="Cannot find endpoint with IP 172.8.8.3" ipAddr="172.8.8.3:36245" subsys=fqdn/dnsproxy
level=error msg="cannot extract endpoint ID from DNS request" DNSRequestID=25703 dnsName=126.1.0.10.in-addr.arpa. error="Cannot find endpoint with IP 172.8.8.3" ipAddr="172.8.8.3:36773" subsys=fqdn/dnsproxy
level=error msg="cannot find matching endpoint" error="DNS request cannot be associated with an existing endpoint" l3n4Addr="172.8.8.3:36773" subsys=daemon
level=error msg="cannot find matching endpoint" error="DNS request cannot be associated with an existing endpoint" l3n4Addr="172.8.8.3:36245" subsys=daemon
...
```

Not sure if there's a connection between Cilium and MetalLB, especially since I can query the webhook using curl on https://webhook-service.metallb-system.svc:443/validate-metallb-io-v1beta1-ipaddresspool?timeout=10s.

---
### MetalLB Version

0.13.12

### Deployment method

Manifests

### Main CNI

cilium

### Kubernetes Version

1.28.2

### Cluster Distribution

kubeadm

**Comment by conradwt** [comment]
I have started seeing the same issue in the last couple of days.  Next, I'm creating the cluster as follows:

```zsh
minikube start -p gateway-api-kong --nodes=3 --kubernetes-version=v1.31.0
```

Next, I'm installing MetalLb by using the following:

```zsh
kubectl apply -f https://raw.githubusercontent.com/metallb/metallb/v0.14.8/config/manifests/metallb-native.yaml
```

Finally, I'm applying the following manifest for the IP pools:

```yaml
apiVersion: metallb.io/v1beta1
kind: IPAddressPool
metadata:
  name: demo-pool
  namespace: metallb-system
spec:
  addresses:
    # replace the following with your IP address range
    - 192.1.2.100-192.1.2.110
```

When I set up things using Cilium, it worked as expected.  BTW, I use the following to setup for Cilium:

https://github.com/conradwt/k8s-gateway-api-using-cilium

**Comment by fedepaol** [comment]
The behaviour of webhooks is generic and do not depend on metallb in any way. It's a service that gets hit by the api server to validate the requests it's configured for.

@CallisteH your test demonstrates that the webhook (or at least the service behind it) is configured correctly. The main difference with when it's executed by the api server is that the apiserver is a host networked pod, so the proper test would be to try to reach the service from a host where teh apiserver is running.

What I saw over time is that the majority of the errors were about having firewalls or proxy settings on the host. The last straw solution is to change the failure policy to ignore. You won't have validation but metallb will work.

It's also worth having a look at  https://github.com/metallb/metallb/issues/1597 and the article https://hackmd.io/@maelvls/debug-cert-manager-webhook

Hope this helps

**Comment by fedepaol** [comment]
Closing this issue as no feedback was provided. Feel free to reopen it in case you have more information or still think it might be a metallb issue. Thanks!
